### DISCUSSION TITLE
Is it possible to use SQLite FTS(Full-Text-Search)5 in SQLAlchemy ORM or Core?

### DISCUSSION THREAD

**Discussion by capymind** [sqlalchemy/sqlalchemy#9466] [Usage Questions] [answered]
Hi!
First of all, thanks for maintaining this wonderful open source project for decades.

Although I have googling about "using SQLite FTS5 with SQLAlchemy", there is no adequate resources about that.
I think FTS5 is very specific to SQLite, so there is no support for it. Am I right...?

The reason why I have question about that is that, for example, `Post` ORM object has columns of `content` and `title` and
I want to search `Post` based on user input. Although there are ways to search using `like` and SQLAlchemy ORM querying,
I am just curious about possibility of applying FTS5 to my orm objects.

I don't try using SQLAlchemy Core, `text()` and other tools yet.

Sorry if this question doesn't make sense. I am not good at programming (and English).

Thanks for reading this question. 
SQLAlchemy is awesome and really helpful for me.

**Comment by CaselIT** [comment, marked as answer]
Hi,

do you have some links with the specification and / or some examples of use of that feature in sql?
if it uses function or custom operation you should be able to use it without requiring changes in sqlalchemy

**Reply by capymind** [reply]
I am sorry for inconvenience casued by my insufficient explanations.

<br>

> do you have some links with the specification

- SQLite FTS5 documenation: https://www.sqlite.org/fts5.html (quite long, :( )

<br>

>  some examples of use of that feature in sql?

- This is my example using SQLite FTS

    - Step1) Create Table (enabled FTS5) and populate some data
    
        ```python
          from sqlalchemy import create_engine
          from sqlalchemy import text
          from sqlalchemy import select, insert
          from sqlalchemy import MetaData, Table
          
          engine = create_engine("sqlite://", echo=True)
          connection = engine.connect()
          connection.execute(text(
              "CREATE VIRTUAL TABLE email USING fts5(sender, title, body)" # 👈 SQLite virtual table using fts5 (SQLite docs)
          ))
          metadata_obj = MetaData()
          email = Table("email", metadata_obj, autoload_with=engine)
          connection.execute(insert(email), [
              dict(sender='zemma', title='hello', body='Hello World!'),
              dict(sender='zemma', title='hello', body='Hello World!'),
              dict(sender='choco', title='zemma', body='Hello World!1'),
              dict(sender='zemma2', title='hello2', body='Hello World!2'),
              dict(sender='zemma3', title='hello3', body='Hello World!3')
          ])
          connection.commit()
        ```
    - Step2) Using "Full Text Search"!
    
       ```python
        connection.execute(text(
            "SELECT * FROM email WHERE email MATCH 'zemma'" # 👈🔥👏 using FTS, email(not column but table) in where condition
        )).all()
        
        # ↓↓↓ Results ↓↓↓
        2023-03-11 09:03:35,782 INFO sqlalchemy.engine.Engine BEGIN (implicit)
        2023-03-11 09:03:35,787 INFO sqlalchemy.engine.Engine SELECT * FROM email WHERE email MATCH 'zemma'
        2023-03-11 09:03:35,790 INFO sqlalchemy.engine.Engine [generated in 0.00863s] ()
        [('zemma', 'hello', 'Hello World!'), # 👈 `sender` column search (at the same time)
         ('zemma', 'hello', 'Hello World!'),
         ('choco', 'zemma', 'Hello World!1')] # 👈 `title` column search (at the same time)
       ```

<br>

@zzzeek said following (below comment)

> the full text operators we have at a dialect-agnostic level are the match operator: https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.match
SQLite uses MATCH so it should work as given

<br>

I don't know about the `match` in SQLAlchemy until now... (thanks!)
However, I think `match` in SQLAlchemy only applied to single column, which is quite different from about raw text SQL query.

```python
connection.execute(select(email).where(email.c.sender.match('zemma'))).all()

# ↓↓↓ Results ↓↓↓
2023-03-11 09:22:26,483 INFO sqlalchemy.engine.Engine SELECT email.sender, email.title, email.body 
FROM email 
WHERE email.sender MATCH ?
2023-03-11 09:22:26,485 INFO sqlalchemy.engine.Engine [cached since 1094s ago] ('zemma',)
[('zemma', 'hello', 'Hello World!'), ('zemma', 'hello', 'Hello World!')]
````

<br>

So, I have to add every individual column condition in `where`.

```python
from sqlalchemy import or_
connection.execute(select(email).where(or_(email.c.sender.match('zemma'), email.c.title.match('zemma')))).all()

# ↓↓↓ Results ↓↓↓
2023-03-11 09:24:25,983 INFO sqlalchemy.engine.Engine SELECT email.sender, email.title, email.body 
FROM email 
WHERE email.sender MATCH ? OR email.title MATCH ?
2023-03-11 09:24:25,987 INFO sqlalchemy.engine.Engine [generated in 0.00358s] ('zemma', 'zemma')
[('zemma', 'hello', 'Hello World!'),
 ('zemma', 'hello', 'Hello World!'),
 ('choco', 'zemma', 'Hello World!1')]
```

<br>

In addition to above basic FTS5 functionality(`WHERE email MATCH 'zemma'`), I think there are many additional features which can be quite useful. The original question is whether I can use this kind of services using SQLAlchemy ORM concepts.. or Core without raw `text()` methods. For example, this is wrong pseudo code..

```python
# this is wrong pseudo code.

Base = DeclarativeBase()
class Email(Base):
  __tablename__ = 'email' # <- sqlite virtual table, which makes FTS5 working (not plain? table)
  sender: Mapped[str]
  title: Mapped[str]
  body: Mapped[str]

session.scalars(select(Email).match('zemma')).all()

```

<br>

Um... example above is not quite useful.. (just using multiple where condition is also okay ..)
However, there are some other features about FTS5 (auxillary `highlight` functions, and `ORDER BY bm25(tablename)` etc...)

Anyway.. I hope that this additional explanation will make previous comment more clear .. :)

Thanks for reading!

**Reply by CaselIT** [reply]
Does this statement actually creates a table? Without specifying any types etc

> `CREATE VIRTUAL TABLE email USING fts5(sender, title, body)`

Or should email already be an existing table?

For the rest I think it's already doable, either using `where(column('email').match(...))` or `where(table.table_valued().match(...))`

**Reply by capymind** [reply]
> Does this statement actually creates a table? 

Yes. The code `CREATE VIRTUAL TABLE email USING fts5(sender, title, body)` actually creates a table.

SQLite docs indicate that

> **It is an error to add types, constraints or PRIMARY KEY declarations to a CREATE VIRTUAL TABLE statement used to create an FTS5 table.** Once created, an FTS5 table may be populated using INSERT, UPDATE or DELETE statements like any other table. Like any other table with no PRIMARY KEY declaration, an FTS5 table has an implicit INTEGER PRIMARY KEY field named rowid.

Your suggestion of using `where(table.table_valued().match(...)` is perfect. Thanks!
On the face of it, that functionality is enough for me to stop considering SQLite FTS5 stuff...

Thanks a lot for the quick and helpful reply.

**Reply by CaselIT** [reply]
To compile that create table, you may use something like
```py
import sqlalchemy as sa
from sqlalchemy.ext.compiler import compiles

engine = sa.create_engine("sqlite://", echo=True)

class FTS5Table(sa.Table):
    pass

@compiles(sa.schema.CreateTable, "sqlite")
def _compile(element: sa.schema.CreateTable, compiler, **kw):
    print("go")
    if not isinstance(element.target, FTS5Table):
        return compiler.visit_create_table(element, **kw)
    name = compiler.preparer.format_table(element.target)
    cols = ", ".join(
        compiler.preparer.format_column(col) for col in element.target.columns
    )
    return f"CREATE VIRTUAL TABLE {name} USING fts5({cols})"

meta = sa.MetaData()
x = sa.Table(
    "t", meta, sa.Column("id", sa.Integer), sa.Column("name", sa.Text)
)
x2 = FTS5Table(
    "t2", meta, sa.Column("id", sa.Integer), sa.Column("name", sa.Text)
)

with engine.connect() as c:
    meta.create_all(c)
```

if you don't like the subclass, just use the table info, doing
```py
    ...
    # if not isinstance(element.target, FTS5Table):
    if not element.target.info.get('use_fts5', False):
    ...

# x2 = FTS5Table(
#     "t2", meta, sa.Column("id", sa.Integer), sa.Column("name", sa.Text)
# )
x2 = sa.Table(
    "t2", meta, sa.Column("id", sa.Integer), sa.Column("name", sa.Text), info={'use_fts5': True}
)
```

**Comment by zzzeek** [comment]
the full text operators we have at a dialect-agnostic level are the match operator: https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Column.match

SQLite uses MATCH so it should work as given

**Reply by capymind** [reply]
I haven't know about `match` operator.

> SQLite uses MATCH so it should work as given

Yes.. it works! 
But is more extensive support (not just to column) of FTS5 possible? (I am not sure that's actually valuable.)
I wrote about that in @CaselIT 's comment above.

Thank you for a quick reply!
